File: src/VitalSynth.Core/HeartOscillator.cs
namespace VitalSynth.Core;

/// <summary>
/// Synthesizes an ECG P-QRS-T complex from a small Fourier series.
/// QRS is reconstructed as a 6-harmonic stack windowed by a super-Gaussian.
/// </summary>
public sealed class HeartOscillator(Cfg cfg)
{
    private const double PStart = 0.10, PEnd = 0.18;
    private const double QStart = 0.22, QEnd = 0.30;
    private const double TStart = 0.38, TEnd = 0.55;

    private const double RPeakPhase = 0.26;

    private readonly Random _rng = new(0xEC6);
    private double _beatPhase;
    private double _rrSec = 60.0 / cfg.Bpm;
    private double _lastBpm = cfg.Bpm;
    private bool _rPeakArmed = true;

    /// <summary>True for exactly one sample when the R-peak of a beat passes.</summary>
    public bool BeatThisSample { get; private set; }

    public int CurrentBpm => cfg.Scenario switch
    {
        Scenario.Asystole => 0,
        Scenario.AFib     => 95 + _rng.Next(-15, 25),
        _                 => (int)Math.Round(_lastBpm),
    };

    public double NextSample(double dt, double respModulation)
    {
        BeatThisSample = false;
        _lastBpm = EffectiveBpm() * (1 + respModulation * cfg.RespDepth);
        _beatPhase += dt / _rrSec;

        if (_beatPhase >= 1)
        {
            _beatPhase -= 1;
            _rPeakArmed = true;
            var jitter = (_rng.NextDouble() - 0.5) * cfg.HrVar / 50;
            _rrSec = 60.0 / Math.Max(_lastBpm, 1) * (1 + jitter);
            if (cfg.Scenario is Scenario.AFib)
                _rrSec *= 1 + (_rng.NextDouble() - 0.5) * 0.35;
        }

        if (_rPeakArmed && _beatPhase >= RPeakPhase && cfg.Scenario is not Scenario.Asystole)
        {
            _rPeakArmed = false;
            BeatThisSample = true;
        }

        if (cfg.Scenario is Scenario.Asystole)
            return GaussianNoise() * cfg.Noise;

        return cfg.QrsMv * (PWave(_beatPhase) + QrsWave(_beatPhase) + TWave(_beatPhase))
             + GaussianNoise() * cfg.Noise;
    }

    private double EffectiveBpm() => cfg.Scenario switch
    {
        Scenario.Tachycardia => Math.Max(cfg.Bpm, 140),
        Scenario.Bradycardia => Math.Min(cfg.Bpm, 42),
        Scenario.AFib        => 95 + 20 * Math.Sin(Environment.TickCount / 4000.0),
        _                    => cfg.Bpm,
    };

    private double GaussianNoise()
    {
        var u1 = 1 - _rng.NextDouble();
        var u2 = 1 - _rng.NextDouble();
        return Math.Sqrt(-2 * Math.Log(u1)) * Math.Cos(2 * Math.PI * u2);
    }

    private static double PWave(double p)
    {
        if (p < PStart || p > PEnd) return 0;
        var x = (p - PStart) / (PEnd - PStart);
        return 0.12 * (0.5 - 0.5 * Math.Cos(2 * Math.PI * x));
    }

    private static double TWave(double p)
    {
        if (p < TStart || p > TEnd) return 0;
        var x = (p - TStart) / (TEnd - TStart);
        return 0.35 * Math.Sin(Math.PI * x);
    }

    private static double QrsWave(double p)
    {
        if (p < QStart || p > QEnd) return 0;
        var x = (p - QStart) / (QEnd - QStart);

        var harmonics = 0.0;
        for (var k = 1; k <= 6; k++)
            harmonics += Math.Sin(k * Math.PI * x) / k;

        var env = Math.Exp(-Math.Pow((x - 0.5) / 0.08, 2));
        var q   = -0.15 * Math.Exp(-Math.Pow((x - 0.25) / 0.05, 2));
        var s   = -0.25 * Math.Exp(-Math.Pow((x - 0.75) / 0.06, 2));
        return env * (0.6 + 0.4 * harmonics) + q + s;
    }
}

File: src/VitalSynth.Core.FS/HeartOscillator.fs
namespace VitalSynth.Core.FS

open System

/// Synthesizes an ECG P-QRS-T complex from a small Fourier series.
///
/// All pattern matches on <see cref="Scenario"/> are exhaustive — listed
/// explicitly without a <c>_</c>-catch-all so the compiler can flag any
/// missing case when a new Scenario is added.
type HeartOscillator(cfg: Cfg) =
    let rng = Random(0xEC6)

    // Internal scratch state stays as raw float — the units are checked
    // at the public boundary in `NextSample` / `CurrentHeartRate`.
    let mutable beatPhase = 0.0
    let mutable rrSec     = 60.0 / float cfg.HeartRate
    let mutable lastBpm   = float cfg.HeartRate
    let mutable rPeakArmed = true
    let mutable beatThisSample = false

    let pStart, pEnd = 0.10, 0.18
    let qStart, qEnd = 0.22, 0.30
    let tStart, tEnd = 0.38, 0.55
    let rPeakPhase   = 0.26

    let gaussianNoise () =
        let u1 = 1.0 - rng.NextDouble()
        let u2 = 1.0 - rng.NextDouble()
        sqrt (-2.0 * log u1) * cos (2.0 * Math.PI * u2)

    let pWave p =
        if p < pStart || p > pEnd then 0.0
        else
            let x = (p - pStart) / (pEnd - pStart)
            0.12 * (0.5 - 0.5 * cos (2.0 * Math.PI * x))

    let tWave p =
        if p < tStart || p > tEnd then 0.0
        else
            let x = (p - tStart) / (tEnd - tStart)
            0.35 * sin (Math.PI * x)

    let qrsWave p =
        if p < qStart || p > qEnd then 0.0
        else
            let x = (p - qStart) / (qEnd - qStart)
            let mutable harmonics = 0.0
            for k in 1 .. 6 do
                harmonics <- harmonics + sin (float k * Math.PI * x) / float k
            let env = exp (-((x - 0.5) / 0.08) ** 2.0)
            let q   = -0.15 * exp (-((x - 0.25) / 0.05) ** 2.0)
            let s   = -0.25 * exp (-((x - 0.75) / 0.06) ** 2.0)
            env * (0.6 + 0.4 * harmonics) + q + s

    /// Exhaustive — adding a new Scenario yields a compiler warning here.
    let effectiveBpm () =
        match cfg.Scenario with
        | Tachycardia       -> max (float cfg.HeartRate) 140.0
        | Bradycardia       -> min (float cfg.HeartRate)  42.0
        | AFib _            -> 95.0 + 20.0 * sin (float Environment.TickCount / 4000.0)
        | Normal
        | Desaturation _
        | Asystole          -> float cfg.HeartRate

    member _.BeatThisSample = beatThisSample

    /// Note the exhaustive match: no `_` wildcard, all six cases listed.
    member _.CurrentHeartRate : float<bpm> =
        match cfg.Scenario with
        | Asystole          -> 0.0<bpm>
        | AFib _            -> (95.0 + float (rng.Next(-15, 25))) * 1.0<bpm>
        | Normal
        | Tachycardia
        | Bradycardia
        | Desaturation _    -> Math.Round(lastBpm) * 1.0<bpm>

    /// Returned in volts — the boundary type carries the unit; callers
    /// project to millivolts via <see cref="Units.inMillivolts"/>.
    member _.NextSample (dt: float<s>, respModulation: float) : float<V> =
        beatThisSample <- false
        lastBpm <- effectiveBpm () * (1.0 + respModulation * cfg.RespDepth)
        beatPhase <- beatPhase + float dt / rrSec

        if beatPhase >= 1.0 then
            beatPhase <- beatPhase - 1.0
            rPeakArmed <- true
            let jitter = (rng.NextDouble() - 0.5) * float cfg.HrJitter / 50.0
            rrSec <- 60.0 / max lastBpm 1.0 * (1.0 + jitter)
            match cfg.Scenario with
            | AFib _ -> rrSec <- rrSec * (1.0 + (rng.NextDouble() - 0.5) * 0.35)
            | Normal | Tachycardia | Bradycardia | Desaturation _ | Asystole -> ()

        if rPeakArmed && beatPhase >= rPeakPhase then
            match cfg.Scenario with
            | Asystole -> ()
            | Normal | Tachycardia | Bradycardia | AFib _ | Desaturation _ ->
                rPeakArmed <- false
                beatThisSample <- true

        let noiseV = gaussianNoise () * float cfg.NoiseAmplitude

        match cfg.Scenario with
        | Asystole -> noiseV * 1.0<V>
        | Normal | Tachycardia | Bradycardia | AFib _ | Desaturation _ ->
            (float cfg.QrsAmplitude * (pWave beatPhase + qrsWave beatPhase + tWave beatPhase)
             + noiseV) * 1.0<V>
